[mlir][linalg] Extend elementwise

javedabsar1 opened the pull request:
Implements Linalg elemwise named-op following the proposal and discussions in RFC:
https://discourse.llvm.org/t/rfc-extend-linalg-elemwise-named-ops-semantics/83927/1
  
I address the comments made in previous PR https://github.com/llvm/llvm-project/pull/122753 but closed that and opened this one as I messed the merge and could not recover.

Just to recap, the main changes I made :
- instead of duplicating Unary/Binary enums to create a new unified one, I have now added some  tablegen functions 'join'/'flatten' to derive unified list. A simple list concat would not work as enums values would clash,
- Fixed tests naming and added more tests as requested.
- Addressed typos/nitpicks.
- changed 'comp_type' to 'kind'  (e.g. add, sub) by popular request.

Some comments that that i did not (could not) do code change:
- default print will not work as it would print default maps.
-  region is needed to lower to linalg.generic.

rengolin commented on mlir/test/Dialect/Linalg/element_wise/invalid.mlir:
```diff
@@ -0,0 +1,54 @@
+// RUN: mlir-opt %s -split-input-file -verify-diagnostics
+func.func @misspelt_op_div(%A : memref<16x8xf32>, %B: memref<16x8xf32>, %C: memref<16x8xf32>) {
+  // expected-error@+3 {{expected ::mlir::linalg::ElementwiseFn to be one of: exp, log, abs, ceil, floor}}
```
Error message is listing unary operations when the op is binary.

javedabsar1 replied:
It will list all the elemwise (13 unary, ~9 binary...). Just this test is keeping the checking short.
Yes 'dive' looks a lot like binary div to our eyes. But error listing will list all potential elemwise.

rengolin commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -3611,5 +3620,259 @@ Speculation::Speculatability MatmulOp::getSpeculatability() {
+} // namespace
+
+static NAryCategoryAndFn getNAryCategoryAndFn(ElementwiseFn fn) {
+  constexpr int lastUnary = static_cast<int>(ElementwiseFn::erf);
```
This seems fragile. If people add past the last, this will break the logic. Perhaps add a `LAST_UNARY`, `LAST_BINARY` and `LAST_TERNARY` and make it illegal for them to exist on the parser?

javedabsar1 replied:
You are right. I should have thought about this fragility. OK, I fixed it in the TableGen itself and now the check does not depend on checking flaky 'erf', but relies on the sizes of the lists.

rengolin commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgEnums.td:
```diff
@@ -55,6 +55,50 @@ def TernaryFn : I32EnumAttr<"TernaryFn", "", [
+// Join two I32EnumAttrCase lists. This joining takes care that the
+// 'int enum values' in the combined list do not overlap. It does this
+// by adding to each element of second list the offset '!size(a)'.
+class JoinTwoI32EnumAttrCaseList< list<I32EnumAttrCase> a,
```
Thinking out loud, perhaps it would be easier to use bit patterns instead of joining enum lists. We won't have more than 20 operations per `Category`, so:
* Unary: `(op | (0xFF << 1))`
* Binary: `(op | (0xFF << 2))`
* Ternary: `(op | (0xFF << 3))`

And set the enums above like: 
* `I32EnumAttrCase<"log", (1 << 1)>`
* `I32EnumAttrCase<"sub", (1 << 2)>`
* `I32EnumAttrCase<"select", (1 << 3)>`

etc?

Then you don't need all the complex sequential logic in the parser/verifiers.

javedabsar1 replied:
No need anymore.

rolfmorel replied:
Thanks for the alternative join-based approach, @javedabsar1 - It's quite impressive what can done with TableGen! (For better or worse, TableGen is a programming language on its own.)

I would still think an approach like @rengolin's would lead to simpler C++. The scheme I have in mind is to just shift the arity 30 bits, e.g.   `I32EnumAttrCase<"abs", 2>` becomes `I32EnumAttrCase<"abs", 2 + (1 << 30)>`, 
`I32EnumAttrCase<"div", 3 + (2 << 30)>`
`I32EnumAttrCase<"select", 0 + (3 << 30)>`. This way the arity can be retrieved by just shifting right 30 bits (e.g. `derivedEnumVal >> 30`) and to obtain the original op code you just do `derivedEnumVal & ((1 << 30) - 1)`.

Three nested `!lfold`s should now suffice to derive all the derived enum cases and `ElementwiseFnLimits` could go.

banach-space replied:
Let me make a pragmatic case for C++ ...

I've not seen much (any?) sophisticated TableGen in dialect definition. C++ might simply be more familiar to folks. I've already revealed my own personal preference 😅

rengolin reviewed (commented):
Some small comments, but overall looking good. Thanks!

javedabsar1 commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:
```diff
@@ -551,6 +551,122 @@ def BroadcastOp : LinalgStructuredBase_Op<"broadcast", [
+  }];
+
+  let arguments = (ins
+      Variadic<AnyType>:$inputs,
```
thanks. changed to AnyShaped.

rolfmorel replied:
Left a comment about this below: the primary reason to allow `AnyType` is so that a scalar operand can be broadcast to the full output shape.

rengolin commented:
Thanks @javedabsar1 this looks good to me now! I'd like @banach-space and @MaheshRavishankar to have a look, as they participated in the design discussions, too. Also, @rolfmorel, @shahidact and @adam-smnk who have been active on the area.

rolfmorel commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgEnums.td:
```diff
@@ -55,6 +55,65 @@ def TernaryFn : I32EnumAttr<"TernaryFn", "", [
+}
+
+// Define an `enum class : i32` to categorise elementwise ops.
+def ElementwiseNAryCategory : I32EnumAttr<"ElementwiseNAryCategory", "", [
```
Couldn't this just be `ElementwiseArity`? The concept of `NAryCategory` is exactly arity, right?

rolfmorel commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:
```diff
@@ -551,6 +551,122 @@ def BroadcastOp : LinalgStructuredBase_Op<"broadcast", [
+    Linalg op form which performs element-wise computation.
+
+    The attribute `kind` describes the operation (e.g. add, exp). The operation
+    kind can be any elementwise nary (e.g. unary, binary) operation.
```
nit: rephrase to use arity instead of nary

rolfmorel commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:
```diff
@@ -551,6 +551,122 @@ def BroadcastOp : LinalgStructuredBase_Op<"broadcast", [
+    The attribute `kind` describes the operation (e.g. add, exp). The operation
+    kind can be any elementwise nary (e.g. unary, binary) operation.
+
+    Affine-maps for operands and result are required to be provided by the user
+    when transpose and/or broadcast is needed on any operand. When a map is not
```
nit: when _a_ transpose ...

banach-space replied:
1. Could you start with the default behaviour (e.g. "By default, all indexing maps are identities.")?
2. Is it OK to specify only one (or two) of the maps? Or is it "either all or nothing"? Please clarify.

rolfmorel commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:
```diff
@@ -551,6 +551,122 @@ def BroadcastOp : LinalgStructuredBase_Op<"broadcast", [
+    User-defined affine-map for operands and result must only be projected
+    permutations with no zero constants.
+
+    For elementwise, iterator-types are always `all parallel`.
```
nit: all `parallel`

banach-space replied:
Also, skip `For elementwise`. This is duplicating info.

rolfmorel commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:
```diff
@@ -551,6 +551,122 @@ def BroadcastOp : LinalgStructuredBase_Op<"broadcast", [
+      (ins "ValueRange":$inputs, "ValueRange":$outputs,
+            CArg<"ArrayRef<NamedAttribute>", "{}">:$attributes),
+      [{
+        buildElementwiseOp($_builder, $_state, std::nullopt, inputs, outputs,
```
Could we inline `buildElementwiseOp` here? As `buildElementwiseOp` is just forwarding args to `buildStructuredOp`, it seems the separate function (not close to any other elementwise op code) is unnecessary.

rolfmorel commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:
```diff
@@ -551,6 +551,122 @@ def BroadcastOp : LinalgStructuredBase_Op<"broadcast", [
+  let extraClassDeclaration = structuredOpsBaseDecls # [{
+      /// Get the nary category enum, e.g. `ElementwiseNAryCategory::Unary`,
+      /// corresponding to the given fn, e.g. `ElementwiseFn::exp`
+      static ElementwiseNAryCategory getNAryCategory(ElementwiseFn fn);
```
If `ElementwiseNAryCategory` just encodes the arity, do we need a separate enum for this? Would just an unsigned int suffice? (The thing we would lose is that the enum encodes/enforce arities are restricted to unary, binary and ternary - maybe that's sufficient reason to keep it.)

rolfmorel commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:
```diff
@@ -551,6 +551,122 @@ def BroadcastOp : LinalgStructuredBase_Op<"broadcast", [
+      static ElementwiseNAryCategory getNAryCategory(ElementwiseFn fn);
+
+      /// Both user-specified and default indexing map will always depend on
+      ///  the current Op instance.
```
nit: double space

Groverkss commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:
```diff
@@ -551,6 +551,122 @@ def BroadcastOp : LinalgStructuredBase_Op<"broadcast", [
+  }];
+
+  let arguments = (ins
+      Variadic<AnyShaped>:$inputs,
```
+1 we should allow that

rolfmorel commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:
```diff
@@ -551,6 +551,122 @@ def BroadcastOp : LinalgStructuredBase_Op<"broadcast", [
+      unsigned getResultRank();
+
+      /// Returns N 'parallel' iterator types where N is rank of result.
+      SmallVector<utils::IteratorType> getIteratorTypesArray();
```
Given the availability of `getResultRank()`, sounds like the definition could be here in the `.td` file.

Groverkss commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:
```diff
@@ -551,6 +551,122 @@ def BroadcastOp : LinalgStructuredBase_Op<"broadcast", [
+      Variadic<AnyShaped>:$inputs,
+      Variadic<AnyShaped>:$outputs,
+      ElementwiseFnAttr:$kind,
+      DefaultValuedOptionalAttr<AffineMapArrayAttr, "{}">:$indexing_maps
```
I'm not sure if this should be DefaultValuedOptionalAttr. "{}" can be invalid in many cases. Instead, we should just have a builder for having a derived default value of the attribute.

rolfmorel commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:
```diff
@@ -551,6 +551,122 @@ def BroadcastOp : LinalgStructuredBase_Op<"broadcast", [
+      SmallVector<utils::IteratorType> getIteratorTypesArray();
+
+      /// The default indexing maps are identities.
+      /// There will be N such maps, where N is the arity of the Op.
```
nit: should "N" maybe be "`arity`"?

rolfmorel replied:
and `numDims` could be `rank`, right?

rolfmorel commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -3611,5 +3620,260 @@ Speculation::Speculatability MatmulOp::getSpeculatability() {
+}
+
+SmallVector<AffineMap>
+ElementwiseOp::getDefaultIndexingMaps(unsigned numMaps, unsigned numDims,
```
nit: consistent arg names w.r.t. declaration

rolfmorel commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -3611,5 +3620,260 @@ Speculation::Speculatability MatmulOp::getSpeculatability() {
+  SmallVector<Attribute, 3> indexingMapsAttr;
+  Attribute mapAttr;
+  if (succeeded(parser.parseOptionalKeyword("indexing_maps"))) {
+    if (parser.parseEqual())
```
As above: can use `||`

Groverkss commented:
> region is needed to lower to linalg.generic

This is not really true. LinalgStructuredInterface exposes `getRegionBuilder` which should be used to lower to linalg.generic

rolfmorel commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -3611,5 +3620,260 @@ Speculation::Speculatability MatmulOp::getSpeculatability() {
+
+  // At this stage of parsing the only way to infer number of region
+  // args is through op kind, as input output tensors are not parsed yet.
+  auto arityAndCategory = getNAryCategoryAndFn(elemwiseFnVal);
```
Hmm - in which sense are arity and category distinct concepts? (If it's just the range restriction, in what sense do we obtain both the arity and category here?)

rolfmorel commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -3611,5 +3620,260 @@ Speculation::Speculatability MatmulOp::getSpeculatability() {
+
+  // Initialize indexingMaps, if not supplied explicitly.
+  if (indexingMapsAttr.empty()) {
+    // We need to infer the `number of indexing maps` needed from the result
```
nit: you (need to) infer the _rank_ of the affine maps from the output type, not the number of maps.

rolfmorel commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -3611,5 +3620,260 @@ Speculation::Speculatability MatmulOp::getSpeculatability() {
+
+  SmallVector<Attribute, 3> indexingMaps = llvm::map_to_vector(
+      ElementwiseOp::getDefaultIndexingMaps(arity + 1, numDims, getContext()),
+      [](AffineMap map) -> Attribute { return AffineMapAttr::get(map); });
```
Instead of turning values into attributes here (which can be quite costly in a couple of respects), how about extracting the AffineMap values associated to `getIndexingMaps` and do an equality check on two non-attribute-based iterators.

rolfmorel commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:
```diff
@@ -551,6 +551,122 @@ def BroadcastOp : LinalgStructuredBase_Op<"broadcast", [
+    permutations with no zero constants.
+
+    For elementwise, iterator-types are always `all parallel`.
+    Iterator-types are needed for constructing the underlying structured op.
```
Agreed on this.

rolfmorel commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:
```diff
@@ -551,6 +551,122 @@ def BroadcastOp : LinalgStructuredBase_Op<"broadcast", [
+    provided, default identity maps are inferred for each operand. The number
+    of dims in each of the identity maps is equal to the rank of the output type.
+    In the case of default indexing map, all input and output shapes must match.
+    User-defined affine-map for operands and result must only be projected
+    permutations with no zero constants.
```
This is a good question! Are there any considerations arising from the "linalg tree" for why the `indexing_maps` should be project permutations?

rolfmorel commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -3611,5 +3620,260 @@ Speculation::Speculatability MatmulOp::getSpeculatability() {
+LogicalResult ElementwiseOp::verify() {
+  // All necessary checks are done either by
+  // - EnumAttr (e.g. unknown operation kind)
+  // - verifyStructuredOpInterface (incorrect map, sizes).
```
I don't understand: this definition is overriding the default implementation of `verify` that does call out to `verifyStructuredOpInterface` - how are the necessary affine_maps checks still performed if we don't invoke `verifyStructuredOpInterface`?

banach-space replied:
I am also quite confused. And suspect that this is the root cause of a bit weird error messages?

rolfmorel commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -3611,5 +3620,260 @@ Speculation::Speculatability MatmulOp::getSpeculatability() {
+
+  if (category == ElementwiseNAryCategory::Unary) {
+    result =
+        helper.buildUnaryFn(categoryAndFn.fn.unaryFn, block.getArgument(0));
```
In case I understand correctly, `categoryAndFn.fn.unaryFn` and `...fn.binaryFn` and `...fn.ternaryFn` are only used in this function. Could we move the logic for retrieving the actual function here and simplify/get rid of `NAryCategoryAndFn`?

rolfmorel commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -3611,5 +3620,260 @@ Speculation::Speculatability MatmulOp::getSpeculatability() {
+        helper.buildTernaryFn(categoryAndFn.fn.ternaryFn, block.getArgument(0),
+                              block.getArgument(1), block.getArgument(2));
+  } else
+    assert(false && "found unhandled category in elemwise print");
```
nit: this function is not about printing

rolfmorel commented on mlir/test/Dialect/Linalg/element_wise/generalize_named_ops.mlir:
```diff
@@ -0,0 +1,157 @@
+func.func @ternary(%A : tensor<32x16xi1>, %B: tensor<8x16x32xf32>, %C : tensor<8x16x32xf32>, %D : tensor<8x16x32xf32>) ->  tensor<8x16x32xf32> {
+  %r = linalg.elementwise
+               kind=#linalg.elementwise_fn<select>
+               indexing_maps = [affine_map<(d0, d1, d2) -> (d2, d1)>, affine_map<(d0, d1, d2) -> (d0, d1, d2)>, affine_map<(d0, d1, d2) -> (d0, d1, d2)>, affine_map<(d0, d1, d2) -> (d0, d1, d2)>]
```
Can we put the `affine_map`s on `linalg.elementwise` on multiple lines? Would be easier to check/read.

rolfmorel commented on mlir/test/Dialect/Linalg/element_wise/round-trip.mlir:
```diff
@@ -0,0 +1,88 @@
+#map = affine_map<(d0, d1, d2, d3, d4) -> (d0, d1, d2, d3, d4)>
+func.func @redundant_maps(%A: tensor<1x2x3x4x5xi32>, %B: tensor<1x2x3x4x5xi32>,
+                          %C: tensor<1x2x3x4x5xi32>) ->  tensor<1x2x3x4x5xi32> {
+  %r = linalg.elementwise   kind=#linalg.elementwise_fn<mul>
```
nit: kind is on own line everywhere else

rolfmorel commented:
> > region is needed to lower to linalg.generic
> 
> This is not really true. LinalgStructuredInterface exposes `getRegionBuilder` which should be used to lower to linalg.generic

Unfornately, this is not case in the world we live in 😢 Have a look at https://github.com/llvm/llvm-project/blob/main/mlir/lib/Dialect/Linalg/Transforms/Generalization.cpp#L71 - that is, the actual region is copied from the named op.

rolfmorel commented:
This is looking good, @javedabsar1 ! I have left a bunch of minor comments. The only one I feel is more significant is on whether we can do with a simpler representation and C++ for functions-and-their-arities.

banach-space commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:
```diff
@@ -551,6 +551,122 @@ def BroadcastOp : LinalgStructuredBase_Op<"broadcast", [
+                   AttrSizedOperandSegments]> {
+  let summary = [{ Performs element-wise operation }];
+  let description = [{
+    Linalg op form which performs element-wise computation.
```
[nit] This is repeating info from the summary, deleteme.

banach-space commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:
```diff
@@ -551,6 +551,122 @@ def BroadcastOp : LinalgStructuredBase_Op<"broadcast", [
+  let description = [{
+    Linalg op form which performs element-wise computation.
+
+    The attribute `kind` describes the operation (e.g. add, exp). The operation
```
[nit]
```suggestion
    The attribute `kind` describes the arithmetic operation to perform. This operation
    can either be unary (e.g. max), binary (e.g. add) or ternary (i.e. select).
```

banach-space commented on mlir/test/Dialect/Linalg/element_wise/generalize_named_ops.mlir:
Everything in the the Linalg test directory uses hyphen (`-`) rather than underscores (`_`) in filenames. Please follow this convention:
* https://github.com/llvm/llvm-project/tree/main/mlir/test/Dialect/Linalg

banach-space commented on mlir/test/Dialect/Linalg/element_wise/generalize_named_ops.mlir:
```diff
@@ -0,0 +1,157 @@
+// RUN: mlir-opt %s -linalg-generalize-named-ops -split-input-file | FileCheck %s
+// CHECK: #[[MAP:.+]] = affine_map<(d0, d1, d2) -> (d0, d1, d2)>
+//
+// CHECK: @unary_exp(%[[A:.+]]: tensor<8x16x32xf32>, %[[B:.+]]: tensor<8x16x32xf32>)
```
Please use `CHECK-LABEL`.

banach-space commented on mlir/test/Dialect/Linalg/element_wise/generalize_named_ops.mlir:
```diff
@@ -0,0 +1,157 @@
+// CHECK:   %[[EXP:.+]] = math.exp %[[A_ARG]] : f32
+// CHECK:   linalg.yield %[[EXP]] : f32
+//
+func.func @unary_exp(%A : tensor<8x16x32xf32>, %B: tensor<8x16x32xf32>) ->  tensor<8x16x32xf32> {
```
There's no clear naming scheme for tests in this file. Here's a suggestion:
* `@arity_kind{_transpose|broadcast_for_op_1}?{_transpose|broadcast_for_op_2}?`

Below I make a suggestion re testing on memrefs.

banach-space commented on mlir/test/Dialect/Linalg/element_wise/generalize_named_ops.mlir:
```diff
@@ -0,0 +1,157 @@
+// RUN: mlir-opt %s -linalg-generalize-named-ops -split-input-file | FileCheck %s
+// CHECK: #[[MAP:.+]] = affine_map<(d0, d1, d2) -> (d0, d1, d2)>
```
[nit] Lets stick to one style within this file.
```suggestion
// CHECK: #[[IDENTITY:.+]] = affine_map<(d0, d1, d2) -> (d0, d1, d2)>
```

banach-space commented on mlir/test/Dialect/Linalg/element_wise/generalize_named_ops.mlir:
```diff
@@ -0,0 +1,157 @@
+// CHECK:   %[[DIV:.+]] = arith.divf %[[A_ARG]], %[[B_ARG]] : f32
+// CHECK:   linalg.yield %[[DIV]] : f32
+//
+func.func @binary_div_on_memrefs(%A : memref<16x8xf32>, %B: memref<16x8xf32>, %C: memref<16x8xf32>) {
+  linalg.elementwise
+               kind=#linalg.elementwise_fn<div>
+               ins(%A, %B: memref<16x8xf32>, memref<16x8xf32>)
+               outs(%C: memref<16x8xf32>)
+  return
+}
```
This is changing two things compared to other tests:
* operation kind (`div`),
* SSA to Value semantics.

I suggest changing one thing at a time. Specifically, split this into two tests:
* `@binary_div` (version on tensors),
* `@binary_div_on_memrefs` (version on memrefs).

banach-space commented on mlir/test/Dialect/Linalg/element_wise/generalize_named_ops.mlir:
```diff
@@ -0,0 +1,157 @@
+//
+func.func @binary_mul_on_tensors(%A : tensor<16x8xf32>, %B: tensor<16x8xf32>, %C: tensor<16x8xf32>) ->  tensor<16x8xf32> {
+  %r = linalg.elementwise
+               kind=#linalg.elementwise_fn<mul>
```
All tests in this file operate on `f32`, hence you end up with `mulf` and `addf`. Please add some integers (it's worth including element type in test function name).

banach-space commented on mlir/test/Dialect/Linalg/element_wise/invalid.mlir:
```diff
@@ -0,0 +1,54 @@
+// RUN: mlir-opt %s -split-input-file -verify-diagnostics
+func.func @misspelt_op_div(%A : memref<16x8xf32>, %B: memref<16x8xf32>, %C: memref<16x8xf32>) {
+  // expected-error@+3 {{expected ::mlir::linalg::ElementwiseFn to be one of: exp, log, abs, ceil, floor}}
```
1. Some errors start with `'linalg.elementwise' op` and some without. Please be consistent.
2. Why is it printing the C++ name (`::mlir::linalg::ElementwiseFn`) rather than some MLIR equivalent?

banach-space commented on mlir/test/Dialect/Linalg/element_wise/invalid.mlir:
```diff
@@ -0,0 +1,54 @@
+// -----
+
+func.func @unary_too_many_args(%A : memref<8x16x32xf32>, %B: memref<8x16x32xf32>, %C:  memref<8x16x32xf32>) {
+  // expected-error@+3 {{custom op 'linalg.elementwise' [parseNamedStructuredOpRegion] ods-gen generated region expects 2 args, got 3}}
```
Where does `custom op` come from? And `parseNamedStructuredOpRegion`? This error is a bit confusing/convoluted.

banach-space commented on mlir/test/Dialect/Linalg/element_wise/round-trip.mlir:
```diff
@@ -0,0 +1,88 @@
+// RUN: mlir-opt %s -split-input-file | FileCheck %s
+//
+// Note - the functions are named @{unary|binary}_{identity|transpose|broadcast|transpose_a|...}_{exp|mul|div|..}
```
Oh, this is almost identical to what I proposed in one of my comments ❤️ Could you use the same scheme throughout this PR? 🙏🏻 :)

banach-space commented on mlir/test/Dialect/Linalg/element_wise/round-trip.mlir:
```diff
@@ -0,0 +1,88 @@
+// CHECK-SAME:       ins(%[[A]] : tensor<?x16xf32>) outs(%[[B]] : tensor<8x16x?xf32>) -> tensor<8x16x?xf32>
+//
+func.func @unary_projection_tanh(%A: tensor<?x16xf32>,
+                                          %B: tensor<8x16x?xf32>) ->  tensor<8x16x?xf32> {
```
[nit] Inconsistent indentation. Same for other examples.

banach-space reviewed (commented):
Lovely patch, @javedabsar1 , thank you! Clearly in line with
* https://discourse.llvm.org/t/rfc-extend-linalg-elemwise-named-ops-semantics/,

and consistent with some other recent changes to Linalg (looking at indexing maps):
* https://github.com/llvm/llvm-project/pull/123618.

I've skimmed through the easier parts and left some nits. 

I have one high level comment regarding test organisation. If we are to introduce a dedicated sub-dir for this Op (`test/Dialect/Linalg/element-wise`), then we should do the same for other Ops. I'd be OK with that and just want to make sure we test everything consistently. @rolfmorel , to me that would mean moving tests for `linalg.contract` to `test/Dialect/Linalg/contract` (*).

(*) I mean the most standard tests, e.g. "invalid.mlir" etc.

rolfmorel commented:
Sorry to bring this up late, but I feel others might also have missed the elephant in the room: 

### How is this new `linalg.elementwise`/`linalg.elemwise` op supposed to relate to the existing `linalg.elemwise_unary` and `linalg.elemwise_binary` ops?

Somehow I completely missed that these ops already exist. TBF, the RFC and PRs seem to have left this topic out as well.

@javedabsar1, given that their existence provides a clear design alternative (i.e. extend them to support indexing maps and call it a day), could you help me/us understand the benefit of going with a new op? In particular:
 - What's the benefit of lumping these ops under just one op? That is, is there a benefit to lumping arities?
 - Is the intention to deprecate  `linalg.elemwise_unary` and `linalg.elemwise_binary`?
   - If not, why not keep the IR representation much closer to the existing ops? E.g. just a `fun` attribute in attr-dict instead of discussing `kind` etc.

rolfmorel commented:
> I have one high level comment regarding test organisation. If we are to introduce a dedicated sub-dir for this Op (`test/Dialect/Linalg/element-wise`), then we should do the same for other Ops. I'd be OK with that and just want to make sure we test everything consistently. @rolfmorel , to me that would mean moving tests for `linalg.contract` to `test/Dialect/Linalg/contract` (*).
> 
> (*) I mean the most standard tests, e.g. "invalid.mlir" etc.

I would be okay with moving tests to dedicated subdirs - at the moment linalg tests seem quite scattered around to me. With the "linalg tree" in mind, rather than having dirs per op, we could have dirs per sub-tree. That is, we could have a subdir for all named ops that perform contractions and a separate subdir for all named ops that operate elementwise. Not sure if that's the best way to go though.

rengolin commented:
> How is this new `linalg.elementwise`/`linalg.elemwise` op supposed to relate to the existing `linalg.elemwise_unary` and `linalg.elemwise_binary` ops?
> Somehow I completely missed that these ops already exist. TBF, the RFC and PRs seem to have left this topic out as well.

We did discuss this in person and some threads in the forum, and there are two schools:

1. We `move` the op from OpDSL to ODS at the same time we add affine map semantics. Syntax and semantics needs to be identical, except for the new affine map. Changing semantics occur later. It's also slower to converge semantics, if not close enough to the goals. This is what @shahidact is doing with the `matmul` family, and I prefer it that way, but this is a personal preference.
2. We `create` a new op in ODS with affine maps, but there are no requirements on keeping the same syntax and semantics. This is easier to iterate upstream because whoever is using the old ones won't be affected until the move to the new ones. It's also confusing because now you have two similarly named and behaving operations upstream. This was championed by @MaheshRavishankar in both element wise and matmul threads.

I will not claim my preference is better, that's why I'm ok with this going as is. But I also didn't click the green button, because I want to make sure there is consensus and other people also prefer this route.

> * What's the benefit of lumping these ops under just one op? That is, is there a benefit to lumping arities?

On a single op, the enum business is complicated. Bit-field-like encoding helps clean up a bit, but it's still ugly. I'd rather see three ops (unary, binary, ternary), but again, that's my personal preference. If there's consensus on a single one, I'm also happy to acquiesce.


> * Is the intention to deprecate  `linalg.elemwise_unary` and `linalg.elemwise_binary`?

Definitely yes! There can be only one. :crossed_swords:

javedabsar1 commented:
Addressed most of review comments (there were a lot so may have missed some). Will re-scan and see what more I need to change. Other options is we land this, so next stage works I can begin and for changes-requested that are not fundamental but I am just not getting what is requested, folks can go in and change it to their exact liking.

javedabsar1 commented:
Addressed most of review comments (there were a lot so may have missed some). Will re-scan and see what more I need to change. Other options is we land this, so next stage works I can begin and for changes-requested that are not fundamental but I am just not getting what is requested, folks can go in and change it to their exact liking.

javedabsar1 commented:
Thanks folks for the review comments (@rolfmorel @rengolin @banach-space @Groverkss).
@rolfmorel  - apologies. I think i accidentally deleted one of your comments.

rengolin commented:
> > * What's the benefit of lumping these ops under just one op? That is, is there a benefit to lumping arities?
> 
> On a single op, the enum business is complicated. Bit-field-like encoding helps clean up a bit, but it's still ugly. I'd rather see three ops (unary, binary, ternary), but again, that's my personal preference. If there's consensus on a single one, I'm also happy to acquiesce.

After speaking with a few people about this, here's my understanding.

### Move `elemwise_unary` / `elemwise_binary`

#### Pros
* Clean cut. No new (temporary) ops initially, no old (deprecated) later
* Simply move to ODS + affine maps, which is a practice we have agreed for other ops

#### Cons
* More boilerplate in TableGen / C++, though the C++ part can be commoned up
* Potential churn on current users, though we had little to none on `matmuls`

### Create `element_wise`, move, deprecate `elemwise_unary` / `elemwise_binary`

#### Pros
* Single implementation reduces boilerplate in TableGen / C++
* Less churn on current users, have time to move at their pace

#### Cons
* Yet another operation that does the same thing as previous others
* Enum business for unary/binary/ternary isn't trivial (future ABI issues?)
* Users may _"never find the time to move"_ and we'll have duplicated ops for a long time

### Analisys

My view is that:
* If we choose to move, we need to reduce the boilerplate in the C++ with a common implementation (static local methods, super-class in table-gen, etc).
* If we choose to create a new one, we **must** have a deadline to remove the old ones, and that needs to be _soon_ (months, not years).

I personally prefer a move, but I want this to be a consensus.

@rolfmorel @shahidact @banach-space @ftynse @MaheshRavishankar @qedawkins

javedabsar1 commented:
> > . If there's consensus on a single one, I'm also happy to acquiesce.
> 
> After speaking with a few people about this, here's my understanding.

@rengolin , we also discussed having a pass that can easily rewrite all old variations (linalg.add/... * and elemwise_unary/... to the 'one new elementwise'. That way there is zero disruption.

shahidact commented:
> From that perspective, I think having both old and new exist for a while is ok and that way there is also zero disruption to users.

This seems to be the right way without disrupting the current users/flows. However, we should also have an understanding of What is the potential disadvantage of having both representations apart from problem of plenty.

banach-space commented:
> > * What's the benefit of lumping these ops under just one op? That is, is there a benefit to lumping arities?
> 
> On a single op, the enum business is complicated. Bit-field-like encoding helps clean up a bit, but it's still ugly. I'd rather see three ops (unary, binary, ternary), but again, that's my personal preference. If there's consensus on a single one, I'm also happy to acquiesce.

I've not seen this thread reach a conclusion, so here's my 2p. I haven't noticed any strong arguments for or against, but it sounds like the implementation would be simpler with separate ops. +1 to separating into three ops.

> ### Move `elemwise_unary` / `elemwise_binary`
> #### Pros
> * Clean cut. No new (temporary) ops initially, no old (deprecated) later
> * Simply move to ODS + affine maps, which is a practice we have agreed for other ops
> 
> #### Cons
> * More boilerplate in TableGen / C++, though the C++ part can be commoned up
> * Potential churn on current users, though we had little to none on `matmuls`
> 
> ### Create `element_wise`, move, deprecate `elemwise_unary` / `elemwise_binary`
> #### Pros
> * Single implementation reduces boilerplate in TableGen / C++
> * Less churn on current users, have time to move at their pace
> 
> #### Cons
> * Yet another operation that does the same thing as previous others
> * Enum business for unary/binary/ternary isn't trivial (future ABI issues?)
> * Users may _"never find the time to move"_ and we'll have duplicated ops for a long time

Hm, both options include `Move `elemwise_unary` / `elemwise_binary`? 😅 

My view on this is that, if possible, we should aim to reduce the disruptions to users. With `tensor.pack` + `tensor.unpack` it was a bit easier - I was moving Ops _verbatim_. Having `tensor.pack`/`tensor.unpack` + `linalg.pack`/`linalg.unpack` would be very confusing. In this case, it's not so clear-cut - we are updating the semantics.

Hm, both options involve Move elemwise_unary / elemwise_binary? 😅

My view is that, if possible, we should minimize disruptions for users. With tensor.pack + tensor.unpack, the transition was easier because I was moving ops verbatim. Having both tensor.pack/tensor.unpack and linalg.pack/linalg.unpack would have been too confusing. In this case, it's less clear-cut, as we're updating the semantics.

**My Suggestion:**
* If we converge toward a single op (`linalg.elementwise`), keep the old ops for a short transition period (e.g., <3 months). This gives users time to migrate smoothly.
* If we go with three separate ops (`linalg.elemwise_{unary|binary|ternary}`), then the only option is a direct replacement of the old definitions.

HTH, 
-Andrzej

MaheshRavishankar commented:
> My view is that:

> If we choose to move, we need to reduce the boilerplate in the C++ with a common implementation (static local methods, super-class in table-gen, etc).
> If we choose to create a new one, we must have a deadline to remove the old ones, and that needs to be soon (months, not years). We also must agree on an ABI for the enum elements between unary/binary/ternary to avoid future ABI issues.
I personally prefer a move, but I want this to be a consensus.

The characterization is a bit unclear cause both said "move". My initial suggestion for this line of work was to create Linalg Ops V2 while keeping current OpDSL unchanged, and once we know V2 is on par with V1, deprecate V1. The separation would be easier for clients of Linalg to understand. But the consensus was that we just replace the OpDSL ops for `linalg.matmul` , etc. directly. This did seemed to have worked with almost no downstream disruptions (so kudos to that). Since we have decided to do an "in-place" update, I would vote for an move as well, i.e drop elementwise_unary, elementwise_binary, elementwise_ternary ops and just replace them with `elementwise` . IMO the former split didnt make much sense to begin with. 

So TL;DR I would vote for a move to be consistent with what is happening with other ops. To allow downstream users to catch up we should probably give a deprecation notice, and hopefully upload a patch that deprecates the old ops and in favor of the new ones so that this can be used by downstream users to navigate the change.

rengolin commented:
Reading all the comments, there seems to be an underlying theme:
* No one is really convinced we need the separation between unary/binary. It also will help with the linalg op tree argument if all element-wise named ops descend from a single op than multiple.
* We don't want duplication, so if we add this (as is), we'll need to remove the old ones soon. Just like the efforts on `matmul` and `pack`, the idea is to replicate behaviour, so any user of the old ops should be able to just use the new ones in-place.

So, I propose the following steps:
1. We merge this PR as-is, which frees us to start the thread on compute type, mixed precision, etc.
2. We create a separate PR ASAP to remove the unary/binary from OpDSL, and publish a PSA to the forum warning of the deprecation. Give it a few weeks and merge.
3. If we find a serious problem with the removal, we can revert (2) independently, without upsetting the new work in (1).

After this is stabilized, we should talk about the other named ops (`add`, `max`, `select`) and how we bring them into ODS/C++ as aliases of the one `element_wise` operation.

rengolin reviewed (approved):
Approving this as agreed above, let's iterate in-tree.

@javedabsar1, can you submit a follow up PR to remove the previous ones from OpDSL and write an RFC/PSA thread so that we warn people what's going on?

javedabsar1 committed:
[mlir][linalg] Extend linalg elementwise

Implements Linalg elemwise named-op following the proposal and discussions in RFC:
https://discourse.llvm.org/t/rfc-extend-linalg-elemwise-named-ops-semantics/83927/1
